### PR TITLE
Device fix

### PR DESCRIPTION
**Pull request by Dsantra92**

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy/functional/backends/jax/device.py [hunk 1]
```diff
@@ -37,10 +38,16 @@ def dev(
     as_native: bool = False,
 ) -> Union[ivy.Device, jaxlib.xla_extension.Device]:
     if isinstance(x, jax.interpreters.partial_eval.DynamicJaxprTracer):
-        return ""
+        logging.warn(
```
[the remark this answers is not in the document]
Comment (Dsantra92) [reply ivy/functional/backends/jax/device.py:41]: Didn't know that. Will update!

File: ivy/functional/backends/jax/device.py [hunk 2]
```diff
@@ -37,10 +38,16 @@ def dev(
...
+            "Cannot get device from DynamicJaxprTracer, returning empty string."
+        )
+        if as_native:  # fallback
```
[the remark this answers is not in the document]
Comment (Dsantra92) [reply ivy/functional/backends/jax/device.py:44]: Julia said there was something failing in demos most likely. Just a workaround.

File: ivy/functional/backends/tensorflow/device.py [hunk 3]
```diff
@@ -43,36 +35,64 @@ def to_device(
...
+def is_native_dev(device: str, /):
+    if isinstance(device, str) and device[0] == "/":
+        dev_in_split = device[1:].split(":")[-2:]
+        if len(dev_in_split) == 2:
+            if dev_in_split[0] in ["CPU", "GPU", "TPU"] and dev_in_split[1].isnumeric():
+                return True
+    return False
+
+
+def _shorten_device(device: str):
```
[the remark this answers is not in the document]
Comment (Dsantra92) [reply ivy/functional/backends/tensorflow/device.py:88]: Sure!

File: ivy/functional/backends/jax/device.py [hunk 4]
```diff
@@ -50,53 +57,70 @@ def dev(
...
+    # doesn't check if device is a valid device
     if isinstance(device, str):
         return ivy.Device(device)
-    if device is None:
-        return None
-    p, dev_id = (device.platform, device.id)
-    if p == "cpu":
-        return ivy.Device(p)
-    return ivy.Device(p + ":" + str(dev_id))
+    if is_native_dev(device):  # no duck typing
```
[the remark this answers is not in the document]
Comment (Dsantra92) [reply ivy/functional/backends/jax/device.py:86]: Not sure I completely understand here. If we raise the exception first, the code will look like:
```python
def as_ivy_dev(device, /):
    # doesn't check if device is a valid device
    if not isinstance(device, str) or is_native_dev(device):
        raise ivy.utils.exceptions.IvyException(
            f"Cannot convert {device} to an ivy device. Expected a "
            f"jaxlib.xla_extension.Device or str, got {type(device)}"
        )
        return ivy.Device(device)
    if isinstance(device, str):
        return ivy.Device(device)
    else:  # no duck typing
        p, dev_id = (device.platform, device.id)
        if p == "cpu" and dev_id == 0:
            return ivy.Device(p)
        return ivy.Device(p + ":" + str(dev_id))
```
Any reason as to why raising the exception first will be helpful?

File: ivy/functional/backends/paddle/device.py [hunk 5]
```diff
@@ -8,71 +8,118 @@
 import ivy
 from ivy.functional.ivy.device import Profiler as BaseProfiler
 from paddle.fluid.libpaddle import Place
-
+from paddle.fluid import core
```
Comment (MahmoudAshraf97) [inline ivy/functional/backends/paddle/device.py:11]: please note that the `paddle.fluid` namespace and API is going to be deprecated soon, I'll try to find alternatives and suggest changes

File: ivy/functional/backends/paddle/device.py [hunk 6]
```diff
@@ -8,71 +8,118 @@
...
-        return paddle.fluid.libpaddle.CUDAPlace()
+    if isinstance(device, str):
+        if "tpu" in device:
+            raise ivy.utils.exceptions.IvyBackendException(
+                "TPU not supported in paddle backend. Consider using ",
+                "TensorFlow or JaX.",
+            )
+        else:
+            # Internal function handles more edge cases and devices
+            return _get_paddle_place(device)
```
Comment (MahmoudAshraf97) [inline ivy/functional/backends/paddle/device.py:100]: It's preferable that we copy this internal function and not import it to avoid breaking changes later as this is not supposed to be directly used, I assume that it could be much simpler after we restrict it to cpu/gpu only as we don't support other devices anyway

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]
note that paddle crashes the kernel when a gpu device is requested that has an index higher than the number of available devices e.g. "gpu:1" on single gpu machines

File: ivy/functional/backends/jax/device.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vedpatwardhan) [reply ivy/functional/backends/jax/device.py:86]: Yep, ig we could instead do a 
```
if not is_native_dev(device):  # no duck typing
    raise ivy.utils.exceptions.IvyException(
            f"Cannot convert {device} to an ivy device. Expected a "
            f"jaxlib.xla_extension.Device or str, got {type(device)}"
        )
p, dev_id = (device.platform, device.id)
if p == "cpu" and dev_id == 0:
    return ivy.Device(p)
return ivy.Device(p + ":" + str(dev_id))
```
but ig it's quite minor so we can keep it as it is 👍

**Comment by vedpatwardhan** [comment]
Regarding the changes @MahmoudAshraf97 suggested, could you please check this [commit](https://github.com/unifyai/ivy/commit/37c2d7a3d18dfc55add5e6886503055df5a88d6f) @Dsantra92 if it fixes the problem for the `paddle` backend? Also, could you please try running a few tests on the GPU for all backends? Would be great to be sure before we merge. Thanks 🙂

**Comment by Dsantra92** [comment]
Thanks @MahmoudAshraf97 and @vedpatwardhan  for you review. Will check back asap!

**Comment by Dsantra92** [comment]
Added some extra tests for the device logics.

File: ivy/functional/backends/jax/device.py [hunk 7]
```diff
@@ -50,53 +57,70 @@ def dev(
...
 def _to_device(x, device=None):
     if device is not None:
-        cur_dev = as_native_dev(dev(x))
+        device = as_native_dev(device)
+        cur_dev = dev(x, as_native=True)
         if cur_dev != device:
-            x = jax.device_put(x, as_native_dev(device))
+            # Uses async dispatch, data commited only when required
+            x = jax.device_put(x, device)
     return x
```
Comment (MahmoudAshraf97) [inline ivy/functional/backends/jax/device.py:79]: this function is redundant as calling `to_device` directly should return native arrays if called correctly, I suggest removing it and using the native `to_device`

File: ivy/functional/backends/numpy/device.py [hunk 8]
```diff
@@ -12,17 +12,31 @@
...
 
-def as_native_dev(device: str, /):
-    return "cpu"
+def as_native_dev(device, /):
+    if isinstance(device, str):
+        dev = ivy.Device(device).split(":")[0]
+        if dev != "cpu":
+            ivy.warn(
+                f"Native Numpy does not support {dev} placement. Using cpu instead."
+                + " This behavious can change in the future."
```
Comment (MahmoudAshraf97) [inline ivy/functional/backends/numpy/device.py:28]:
```suggestion
                + " This behaviour can change in the future."
```

File: ivy/functional/backends/numpy/device.py [hunk 9]
```diff
@@ -12,17 +12,31 @@
 
 
 def dev(x: np.ndarray, /, *, as_native: bool = False) -> Union[ivy.Device, str]:
-    if as_native:
-        return "cpu"
```
Comment (MahmoudAshraf97) [inline ivy/functional/backends/numpy/device.py:16]: I know numpy doesn't have native devices, but `NativeDevice` in numpy backend is defined as `str` so I guess it should be returning something?

File: ivy/functional/backends/numpy/device.py [hunk 10]
```diff
@@ -68,21 +82,8 @@ def to_device(
...
-        elif "cpu" in device:
-            pass
-        else:
-            raise ivy.utils.exceptions.IvyException(
-                "Invalid device specified, must be in the form "
-                "[ 'cpu:idx' | 'gpu:idx' ], but found {}".format(device)
-            )
-    return x
+    # TODO: implement stream and out
+    return _to_device(x, device)
```
Comment (MahmoudAshraf97) [inline ivy/functional/backends/numpy/device.py:86]: the same comment about jax `to_device`

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.
